src/out_of_order_engine.cc

```diff
@@ -0,0 +1,414 @@
+	bool probe_for_eager_assignment = true;
+};
+
+/// Instruction is inserted into `assignment_queue` after being popped might be assignable by eager enqueueing to the lane where all its remaining predecessors
+/// are currently executing on, provided no submission of a third instruction is made to the same lane in the meantime.
+/// Implies `num_unassigned_predecessors == 0`.
+struct eagerly_assignable_state {
+	std::optional<device_id> device;
+	lane_id lane = -1;
+	const instruction* expected_last_submission_on_lane = nullptr;
+};
+
+/// Instruction is inserted into `assignment_queue` and is immediately assignable after being popped.
+/// Implies `num_incomplete_predecessors == 0 && num_unassigned_predecessors == 0`.
+struct immediately_assignable_state {};
+
+/// Instruction is assigned and waiting for a `complete_assigned()` call. Implies `num_incomplete_predecessors == 0 && num_unassigned_predecessors == 0`.
+struct assigned_state {
+	std::optional<device_id> device;
+	std::optional<lane_id> lane;
+};
+
+/// State maintained for every instruction between its submission and completion.
+struct incomplete_instruction_state {
+	const instruction* instr = nullptr;
+	out_of_order_engine::target target = out_of_order_engine::target::immediate;
+	gch::small_vector<device_id, 2> eligible_devices; ///< device-to-device copies can be submitted to host or destination device.
+	gch::small_vector<instruction_id> successors;     ///< we collect successors as they are submitted
+
+	/// An instruction with no incomplete predecessors is ready for immediate assignment.
+	size_t num_incomplete_predecessors = 0;
+
+	/// An instruction with no unassigned (but some incomplete) predecessors may be eligible for eager assignment (see try_mark_for_assignment).
+	size_t num_unassigned_predecessors = 0;
+
+	/// Data that only exist in specific assignment states.
+	std::variant<unassigned_state, eagerly_assignable_state, immediately_assignable_state, assigned_state> assignment;
+
+	explicit incomplete_instruction_state(const instruction* instr) : instr(instr), assignment(unassigned_state()) {}
+};
+
+/// State maintained per host thread queue or device in-order queue.
+struct lane_state {
+	size_t num_in_flight_assigned_instructions = 0;
+	const instruction* last_incomplete_submission = nullptr;
+};
+
+/// State maintained for every "multi-lane" target, i.e., `target::host_queue` and `target::device_queue[num_devices]`.
+struct target_state {
+	std::vector<lane_state> lanes;
+};
+
+// Implementation is behind a pimpl because we expect it to grow in complexity when adding support for waiting on inter-queue events to increase scheduling
+// "eagerness" when an instruction needs to wait on multiple incomplete predecessors. This can already be implemented with AdaptiveCpp using the
+// enqueue_custom_operation extension, but requires a similar interop story in DPC++, see https://github.com/intel/llvm/issues/13706 .
+struct engine_impl {
+	system_info system; // immutable
```

Comment:
Why not `const`?

Reply:
I try to follow [this Guideline](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-constref) in all of my code

Reply:
You got me there. But if I make it const clang-tidy will complain for the same exact reason :P

Reply:
Does it complain if you explicitly make it non-copyable?

Reply:
It doesn't!